Add performance profiling to flare command

bmikaili opened the pull request:
### What does this PR do?

Adds support for a heap profile and CPU profile to be taken and included in a flare. The duration of the CPU profile can be configured (30s +). Two heap profiles are taken to allow for diff viewing. The Flare process gets the profiles from the expvar endpoint of the agent, encodes them, and sends them over HTTP for the agent to include in the flare. If there's no profile sent, the agent simply doesn't include it.

A prior draft was https://github.com/DataDog/datadog-agent/pull/6016, but this implementation makes less changes for the same outcome so it is preferred.

### Motivation

It would be useful to have performance profiling data for some support cases.

### Describe your test plan

1. Test that flare without the flag doesn't include the profiles.
2. Test that flare returns error when doing profile with a parameter below 30.
3. Test that files are included and `go tool pprof` works on them when profile flag is set with a value above 30.
Extra for QA: Test that flare isn't broken on other paths (windows, GUI).

Commit:
Added logic for local flare creation

Commit:
Added logic for IPC flare creation and configurable profiling time

Commit:
Add logic to change profiling seconds when flare is reexecuted

Commit:
added reno

Commit:
Removed debug comment

Commit:
Fixed export comment

ogaca-dd commented on pkg/flare/archive.go:
```diff
@@ -746,6 +769,41 @@ func zipHTTPCallContent(tempDir, hostname, filename, url string) error {
+func zipPerformanceProfile(tempDir, hostname, profileDir string) error {
+	srcHeapPath := filepath.Join(profileDir, HeapProfileName)
+	srcHeapProfile, err := os.OpenFile(srcHeapPath, os.O_RDWR|os.O_CREATE, os.ModePerm)
+	if err != nil {
```
Missing `defer srcHeapProfile.Close()`

bmikaili replied:
Added

ogaca-dd commented on pkg/flare/archive.go:
```diff
@@ -746,6 +769,41 @@ func zipHTTPCallContent(tempDir, hostname, filename, url string) error {
+
+	_, err = io.Copy(dstCPUProfile, srcCPUProfile)
+
+	return err
```
There is a`CopyFile` function: https://github.com/DataDog/datadog-agent/blob/18d5c0811f2a9de88f2f48d77eaae19b8c04eea4/pkg/util/common.go#L30.

bmikaili replied:
Ah that's nice, didn't know that.

ogaca-dd reviewed (commented):
LGTM, just few comments.

bmikaili commented on cmd/agent/api/agent/agent.go:
```diff
@@ -80,13 +82,20 @@ func getHostname(w http.ResponseWriter, r *http.Request) {
+	profileDir, ok := vars["profileDir"]
+
+	if ok {
+		profileDir = filepath.Join("/tmp", profileDir)
```
Yeah you are right. I would use an absolute path if I knew how to do it with `mux`.

bmikaili commented on cmd/agent/api/agent/agent.go:
```diff
@@ -80,13 +82,20 @@ func getHostname(w http.ResponseWriter, r *http.Request) {
 
 func makeFlare(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	profileDir, ok := vars["profileDir"]
```
I tried doing that, the problem is, that my url looks like that `https://localhost:5001/agent/flare//tmp/e4f171debf982e3e5692025137839` Which I think is messing with the routing

bmikaili commented on cmd/agent/app/flare.go:
```diff
@@ -116,22 +143,23 @@ func makeFlare(caseID string) error {
 	}
-	urlstr := fmt.Sprintf("https://%v:%v/agent/flare", ipcAddress, config.Datadog.GetInt("cmd_port"))
+
+	urlstr := fmt.Sprintf("https://%v:%v/agent/flare%v", ipcAddress, config.Datadog.GetInt("cmd_port"), profileDir)
```
I see, I didn't know how to pass the path properly in the URL. I will try it with query strings.

bmikaili commented on cmd/agent/app/flare.go:
```diff
@@ -141,17 +169,58 @@ func requestArchive(logFile string) (string, error) {
+		return err
+	}
+
+	err = writeHTTPCallContent(profileDir, flare.HeapProfileName, heapProfURL)
```
In `writeHTTPCallContent` I use the `os.APPEND` flag so it doesn't override it I think.

bmikaili commented on pkg/flare/archive.go:
```diff
@@ -84,19 +89,41 @@ type filePermsInfo struct {
+	}
+
+	dirName := hex.EncodeToString(b)
+	return ioutil.TempDir("", dirName)
```
I am not sure why this was originally done with a random hex. I am just reusing the function. There might have been a reason to do it like that?

bmikaili commented on cmd/agent/app/flare.go:
```diff
@@ -32,6 +41,7 @@ func init() {
 	flareCmd.Flags().StringVarP(&customerEmail, "email", "e", "", "Your email")
 	flareCmd.Flags().BoolVarP(&autoconfirm, "send", "s", false, "Automatically send flare (don't prompt for confirmation)")
 	flareCmd.Flags().BoolVarP(&forceLocal, "local", "l", false, "Force the creation of the flare by the command line instead of the agent process (useful when running in a containerized env)")
+	flareCmd.Flags().IntVarP(&profiling, "profile", "p", 0, "Add performance profiling data to the flare. Will collect the CPU profile for the configured amount of seconds, with a minimum of 30s")
```
Will add.

Commit:
better logging, simplified copying of profiles, absolute path passed in query

L3n41c committed:
Update the shellcheck download link (#6064)

to avoid the following error:
```
You are downloading ShellCheck from an outdated URL!

Please update to the new URL:
https://github.com/koalaman/shellcheck/releases/download/v0.7.0/shellcheck-v0.7.0.linux.x86_64.tar.xz

For more information, see:
https://github.com/koalaman/shellcheck/issues/1871

PS: Sorry for breaking your build. The hosting costs were getting out of hand :(

Exited with code exit status 1
```

ogaca-dd commented on cmd/agent/api/agent/agent.go:
```diff
@@ -40,7 +40,7 @@ import (
 	r.HandleFunc("/version", common.GetVersion).Methods("GET")
 	r.HandleFunc("/hostname", getHostname).Methods("GET")
-	r.HandleFunc("/flare", makeFlare).Methods("POST")
+	r.HandleFunc("/flare", makeFlare).Methods("POST").Queries("profileDir", "{path:.*}")
```
I am wondering if it coud lead to a security issue as this code returns a copy of the files in `profileDir` with the access rights of the running agent.

bmikaili replied:
Technically it just returns the path to the temporary files in `profileDir`, I'm not sure if that's a security issue 🤔

Commit:
Refactored logic to send profile files over HTTP

Commit:
Simplified logic for request to env var endpoint

Commit:
make createtempdir private again because it is unused

Commit:
update createtempdir to private

gbbr commented on cmd/agent/app/flare.go:
```diff
@@ -45,6 +52,8 @@ var flareCmd = &cobra.Command{
 			color.NoColor = true
 		}
 
+		cpuProfURL = fmt.Sprintf("http://127.0.0.1:%s/debug/pprof/profile?seconds=%d", config.Datadog.GetString("expvar_port"), profiling)
```
Would it be possible to get the `trace-agent` profiles too? The mechanism and paths are exactly identical, the only thing that would change is the port.

bmikaili replied:
Yeah would definitely be possible. Might have to leave it to a different PR though because I'm not sure I will get to adding that before friday, and I want to merge this before the freeze.

gbbr commented on cmd/agent/app/flare.go:
```diff
@@ -116,42 +137,75 @@ func makeFlare(caseID string) error {
 
-func requestArchive(logFile string) (string, error) {
+func requestArchive(logFile string, profile *flare.Profile) (string, error) {
 	fmt.Fprintln(color.Output, color.BlueString("Asking the agent to build the flare archive."))
```
IMHO these "codecov" checks are making reviews quite difficult and are only helpful to the author. They could easily remain as part of the CI.

gbbr reviewed (commented):
Sorry for coming late to the party, I might miss some context.

Commit:
trigger pipeline

hush-hush commented on cmd/agent/app/flare.go:
```diff
@@ -45,6 +52,8 @@ var flareCmd = &cobra.Command{
 			color.NoColor = true
 		}
 
+		cpuProfURL = fmt.Sprintf("http://127.0.0.1:%s/debug/pprof/profile?seconds=%d", config.Datadog.GetString("expvar_port"), profiling)
```
The raw URL should be move to a `const` block along with the one for memory.

bmikaili replied:
Will that update the URL with the value for the profiling seconds though?

bmikaili replied:
If I make them consts, I can't put in the value from the config for the port either with `fmt.Sprintf`

bmikaili replied:
Nvm, I moved these to `CreatePerformanceProfile` and moved the method to the `flare` package.

hush-hush commented on cmd/agent/app/flare.go:
```diff
@@ -116,42 +137,75 @@ func makeFlare(caseID string) error {
 	return filePath, nil
 }
+
+func createPerformanceProfile() (*flare.Profile, error) {
+	// Two heap profiles for diff
+	c := util.GetClient(false)
+	firstHeapProf, err := util.DoGet(c, heapProfURL)
+	if err != nil {
+		return nil, err
+	}
+
+	cpuProf, err := util.DoGet(c, cpuProfURL)
+	if err != nil {
+		return nil, err
+	}
+
+	secondHeapProf, err := util.DoGet(c, heapProfURL)
+	if err != nil {
+		return nil, err
+	}
+
+	return &flare.Profile{
+		FirstHeapProfile:  firstHeapProf,
+		CPUProfile:        cpuProf,
+		SecondHeapProfile: secondHeapProf,
+	}, nil
+}
```
This should be moved to the `flare` package as a helper.

hush-hush commented on cmd/agent/app/flare.go:
```diff
@@ -81,12 +90,24 @@ func makeFlare(caseID string) error {
 	var err error
+	if profiling >= 30 {
+		fmt.Fprintln(color.Output, color.BlueString("Creating a %ds performance profile.", profiling))
+		profile, err = createPerformanceProfile()
```
Should we not use `profiling` somewhere to pull profiling seconds of data ?

bmikaili replied:
Yes, in the `cpuProfURL`

hush-hush commented on cmd/agent/app/flare.go:
```diff
@@ -116,42 +137,75 @@ func makeFlare(caseID string) error {
+	b := bytes.NewBuffer([]byte{})
+	if profile != nil {
+		if err := json.NewEncoder(b).Encode(profile); err != nil {
+			fmt.Fprintln(color.Output, color.RedString(fmt.Sprintf("Error while encoding profile: %s", e)))
```
We should stop and exit if we can't serialize the profile. We don't want to send half baked flare.

hush-hush commented on pkg/flare/archive.go:
```diff
@@ -35,10 +35,13 @@ import (
 )
 
 const (
+	heapProfileName = "heap_profile.log"
```
This is also a `.profile ` and not a log

bmikaili replied:
Made it a `.log` because `go-routine-dump` is one as well, but yes, it would be more readable as `.profile`. I will change it.

hush-hush commented on pkg/flare/archive.go:
```diff
@@ -84,18 +87,25 @@ type filePermsInfo struct {
 	group string
 }
 
+// Profile is a performance profile containing heap and CPU profiles.
+type Profile struct {
+	FirstHeapProfile  []byte
+	CPUProfile        []byte
+	SecondHeapProfile []byte
+}
+
 // CreateArchive packages up the files
-func CreateArchive(local bool, distPath, pyChecksPath, logFilePath string) (string, error) {
+func CreateArchive(local bool, distPath, pyChecksPath, logFilePath string, profile *Profile) (string, error) {
 	zipFilePath := getArchivePath()
 	confSearchPaths := SearchPaths{
 		"":        config.Datadog.GetString("confd_path"),
 		"dist":    filepath.Join(distPath, "conf.d"),
 		"checksd": pyChecksPath,
 	}
-	return createArchive(zipFilePath, local, confSearchPaths, logFilePath)
+	return createArchive(confSearchPaths, local, zipFilePath, logFilePath, profile)
 }
 
-func createArchive(zipFilePath string, local bool, confSearchPaths SearchPaths, logFilePath string) (string, error) {
+func createArchive(confSearchPaths SearchPaths, local bool, zipFilePath, logFilePath string, profile *Profile) (string, error) {
```
Why switch argument order ?

bmikaili replied:
Because then I only have to write `string` once. It's a minor change but I thought it's nicer.

truthbk replied:
I agree that it's a nicer interface, it's an unexported function and the likelihood of a regression is small as we're covered by tests and the compiler would fail to build given the change. I'm fine with this.

hush-hush commented on pkg/flare/archive.go:
```diff
@@ -746,6 +762,33 @@ func zipHTTPCallContent(tempDir, hostname, filename, url string) error {
+	if err != nil {
+		return err
+	}
+	_, err = heapProfile.Write(profile.FirstHeapProfile)
+	if err != nil {
+		return err
+	}
+	_, err = heapProfile.Write(profile.SecondHeapProfile)
+	return err
```
Does writing both profile to the same file won't break `pprof` tools ? Should we not write them to 2 different files ?

bmikaili replied:
Yes. Will change together with the file suffix.

hush-hush commented on pkg/flare/archive.go:
```diff
@@ -84,18 +87,25 @@ type filePermsInfo struct {
 
+// Profile is a performance profile containing heap and CPU profiles.
+type Profile struct {
+	FirstHeapProfile  []byte
+	CPUProfile        []byte
+	SecondHeapProfile []byte
```
Should we store a slice of `[]byte` for the heap ? Do we have any use cases where we might want 3 or more heap profile ?

small nit: if we keep the 2 profile could we put them next to each other in the struct:

```go
type Profile struct {
￼	CPUProfile        []byte
￼	FirstHeapProfile  []byte
￼	SecondHeapProfile []byte
}
```

bmikaili replied:
It's only 2 profiles because one is created before the cpu profile, and one after the cpu profile to have a diff. Don't think we need more than that.

Next to each other makes sense 👍

hush-hush commented on cmd/agent/api/agent/agent.go:
```diff
@@ -80,13 +81,23 @@ func getHostname(w http.ResponseWriter, r *http.Request) {
+	case err == io.EOF:
+		log.Debugf("No profile, continuing.")
+	case err != nil:
+		log.Errorf("Error while decoding profile from request body: %s", err)
```
We should return an error here instead of sending the flare.

bmikaili replied:
You mean in the second case?

hush-hush replied:
yes, if the json decode fails

hush-hush commented on cmd/agent/api/agent/agent.go:
```diff
@@ -80,13 +81,23 @@ func getHostname(w http.ResponseWriter, r *http.Request) {
+
+	err := json.NewDecoder(r.Body).Decode(&profile)
+	switch {
+	case err == io.EOF:
+		log.Debugf("No profile, continuing.")
```
Would it not be simpler to check the size of `r.Body` first instead of starting a new Json decoder ?

bmikaili replied:
It would be, I did it this way due to the reasoning in [this](https://stackoverflow.com/a/32718077) stackoverflow comment. Do you think we can disregard that as it is just internal communication?

hush-hush replied:
I was thinking about something like that:

```go
  if r.Body != "" {
	if err := json.NewDecoder(r.Body).Decode(&profile); err != nil {
		// handle error
	}
  }
```

I don't think we need to instantiate and run the JSON decoder if the body is empty.

bmikaili replied:
That makes sense to me, didn't think about doing it that way 👍

bmikaili commented on cmd/agent/app/flare.go:
```diff
@@ -32,6 +38,7 @@ func init() {
 	flareCmd.Flags().StringVarP(&customerEmail, "email", "e", "", "Your email")
 	flareCmd.Flags().BoolVarP(&autoconfirm, "send", "s", false, "Automatically send flare (don't prompt for confirmation)")
 	flareCmd.Flags().BoolVarP(&forceLocal, "local", "l", false, "Force the creation of the flare by the command line instead of the agent process (useful when running in a containerized env)")
+	flareCmd.Flags().IntVarP(&profiling, "profile", "p", 0, "Add performance profiling data to the flare. Will collect the CPU profile for the configured amount of seconds, with a minimum of 30s")
```
This actually enables both, I just need to mention it in the text.

bmikaili commented on cmd/agent/app/flare.go:
```diff
@@ -81,12 +90,24 @@ func makeFlare(caseID string) error {
+			return err
+		}
+	} else {
+		fmt.Fprintln(color.Output, color.YellowString("Profiling not configured or below 30 seconds, skipping it."))
```
The problem with exiting with error is that by default it is set as an invalid time range (0), because profiling is disabled by default. Should I change the default for this?

bmikaili replied:
Nvm I'll just make the default -1 lol

bmikaili commented on cmd/agent/app/flare.go:
```diff
@@ -7,23 +7,29 @@ package app
 
 var (
+	cpuProfURL  string
+	heapProfURL = fmt.Sprintf("http://127.0.0.1:%s/debug/pprof/heap?debug=2",
+		config.Datadog.GetString("expvar_port"))
```
I see. This happens in a bunch of places though, for example in `archive.go`. I'll refactor those too then?

Commit:
Implement maxime's feedback

Commit:
Only decode body for profile if body is not empty

truthbk commented on pkg/flare/archive.go:
```diff
@@ -84,18 +88,54 @@ type filePermsInfo struct {
+
+// CreatePerformanceProfile creates a CPU and Heap profile to include
+func CreatePerformanceProfile(profileDuration int) (*Profile, error) {
+	cpuProfURL := fmt.Sprintf("http://127.0.0.1:%s/debug/pprof/profile?seconds=%d", config.Datadog.GetString("expvar_port"), profileDuration)
```
Though this might be out of scope, I feel like it's only right to bring it up. Should we add facilities to collect the relevant profiles from other agents (trace, process,...) exposing the `pprof` API?  It shouldn't be particularly difficult to at least lay the foundation to perform such collection. For instance, we could make the cpuProfURL parametric here:

```golang
// caller is responsible for providing a valid baseURL ie. fmt.Sprintf("http://127.0.0.1:%s/debug/pprof/profile", config.Datadog.GetString("expvar_port")
// but could also be remote, or an alternate port on a different agent.
func CreatePerformanceProfile(baseURL string, profileDuration int) (*Profile, error) {
	cpuProfURL := fmt.Sprintf("%s?seconds=%d", baseURL, profileDuration)
    // ...
}

```

We might also want to consider an alternative where the profiles are submitted via the API in a push model rather than a pull model. Now this is would 100% be out of scope :)

And of course all of the above can be tackled later. So feel free to ignore.

bmikaili replied:
The first one is a quick change and makes sense to me, I'll change it.

The second one would be neat to have, but yeah that one's out of scope. Should I add a card in the backlog or document it elsewhere?

truthbk replied:
@bmikaili I'll create a card - definitely low priority at the moment.

truthbk reviewed (commented):
Mostly looks good to me after addressing @hush-hush changes, just added a small question so we can maybe reuse some code in future use-cases.

Commit:
changed CreatePerformanceProfile to accept profile base URL as parameter for future profile additions

mx-psi reviewed (approved):
Approving; the codeowner request was only because of the shellcheck fix

hush-hush commented on cmd/agent/app/flare.go:
```diff
@@ -116,39 +134,48 @@ func makeFlare(caseID string) error {
+
+	b := bytes.NewBuffer([]byte{})
+	if profile != nil {
+		if err := json.NewEncoder(b).Encode(profile); err != nil {
```
Why not simply use `json.Marchall(profile, b)` ?

bmikaili replied:
In what way? Encode can write to that buffer in less lines.

hush-hush commented on pkg/flare/archive.go:
```diff
@@ -746,6 +792,40 @@ func zipHTTPCallContent(tempDir, hostname, filename, url string) error {
 
+func zipPerformanceProfile(tempDir, hostname string, profile *Profile) error {
+	cpuProfilePath := filepath.Join(tempDir, hostname, cpuProfileName)
+	cpuProfile, err := os.OpenFile(cpuProfilePath, os.O_RDWR|os.O_CREATE, os.ModePerm)
```
Let's call `ensureParentDirsExist(cpuProfile)`, just in case,

hush-hush commented on tasks/test.py:
```diff
@@ -592,7 +592,7 @@ def install_shellcheck(ctx, version="0.7.0", destination="/usr/local/bin"):
 
     ctx.run(
-        "wget -qO- \"https://storage.googleapis.com/shellcheck/shellcheck-v{sc_version}.{platform}.x86_64.tar.xz\" | tar -xJv -C /tmp".format(
+        "wget -qO- \"https://github.com/koalaman/shellcheck/releases/download/v{sc_version}/shellcheck-v{sc_version}.{platform}.x86_64.tar.xz\" | tar -xJv -C /tmp".format(
```
Nit: for next time this should be in a different PR.

bmikaili replied:
It's just cherrypick from master because circleci was failing.

truthbk replied:
In this case my advice would be to:
```
# assuming we're in your-branch
git checkout master && git pull
git checkout your-branch
git rebase master
```

or alternatively:
```
# assuming we're in your-branch
git merge origin master
```

hush-hush replied:
Yes, `rebase` is your best friend.

Commit:
implement maximes feedback

truthbk commented on pkg/flare/archive_test.go:
```diff
@@ -32,7 +32,7 @@ func TestCreateArchive(t *testing.T) {
 	mockConfig.Set("confd_path", "./test/confd")
```
We haven't added a test case here, it probably makes sense - you should be able to just use dummy data, even an empty `profile *Profile` to somewhat easily test the functionality.

bmikaili replied:
Makes sense to me! I will add tests for `zipPerformanceProfile`. Do you think it also makes sense to unit test the method `CreatePerformanceProfile`? That one does a request against the profile endpoints and creates the `profile` struct from it. I would say that one does not make sense to be tested because it would be completely mocked and it would just be a test of our `apiutil`.

truthbk reviewed (commented):
Almost there! Let's add a small test and get this merged tomorrow!

hush-hush commented on cmd/agent/api/agent/agent.go:
```diff
@@ -80,13 +81,22 @@ func getHostname(w http.ResponseWriter, r *http.Request) {
+	var profile *flare.Profile
+
+	if r.Body != http.NoBody {
+		if err := json.NewDecoder(r.Body).Decode(&profile); err != nil {
```
let's use `json.Unmarshal` here too

hush-hush commented on pkg/flare/archive.go:
```diff
@@ -746,6 +792,45 @@ func zipHTTPCallContent(tempDir, hostname, filename, url string) error {
+		return err
+	}
+
+	if err := ensureParentDirsExist(cpuProfilePath); err != nil {
```
You opened the file line 797, we already failed if the folder doesn't exists.

Commit:
add unit test, change to json.unmarshal, change to correct position for ensureparentdirsexist

Commit:
Merge remote-tracking branch 'origin' into berzan/flare-performance-profiling
